utils.groovy

```diff
@@ -29,9 +39,36 @@ def load_pipecfg() {
         ])
         pipecfg = readYaml(file: "pipecfg/config.yaml")
     }
+    validate_pipecfg(pipecfg)
     return pipecfg
 }
 
+def validate_pipecfg(pipecfg) {
+    if (env.JENKINS_URL in PROTECTED_JENKINSES) {
+        // We're running on a protected Jenkins instance. Check the hotfix
+        // build policy and that S3 buckets are as expected.
+        def (bucket_prefix, allow_hotfixes) = PROTECTED_JENKINSES[env.JENKINS_URL]
+        if (pipecfg.hotfix && !allow_hotfixes) {
+            error("Hotfix builds are not allowed on ${env.JENKINS_URL}")
+        }
+        if (pipecfg.s3_bucket && bucket_prefix != pipecfg.s3_bucket) {
+            error("S3 bucket on ${env.JENKINS_URL} must be ${bucket_prefix} (found ${pipecfg.s3_bucket})")
+        }
+    } else if (AWSBuildUploadCredentialExists()) {
+        // We're *not* running on a protected Jenkins instance and we have AWS
+        // creds available. Check that we're not trying to push to a protected
+        // S3 bucket.
+        def protected_buckets = PROTECTED_JENKINSES.collectEntries{jenkins_url, v ->
+            def (bucket_prefix, allow_hotfixes) = v
+            [splitBucketKey(bucket_prefix)[0], jenkins_url]
```

Comment:
What is this code doing? I don't quite understand it.

Reply:
I added a comment. Does that help?

Reply:
Yeah I mean I get what you are doing. I just don't really understand how the code achieves the goal. The `[bucket, jenkins_url]` on a line by itself confuses me.

I'm not too worried about it, though.